### ISSUE TITLE
XML Namespaces

### ISSUE THREAD

**Issue by ballsteve** [ballsteve/xrust#78]
For certain transformations, in particular the system-property XSLT function, the transform engine needs to be able to expand QNames, e.g. `xsl:version` becomes `Q{http://www.w3.org/1999/XSL/Transform}version`. To do that, the transform engine needs to have a copy of the namespaces declared in the stylesheet (not the source document). Since the transform engine is separate to the stylesheet, or can be created without a stylesheet, I've added a mechanism to the parser that includes a copy of the namespace declarations along with the constructed tree - see parse_with_ns (branch feb2-2024).

This works, except that in the parser/xml/element module, when the code checks the namespace of the element it pops the namespace vector. This means that the namespace declaration is lost when the state is returned to the caller.

I've changed the code so that the namespace vector's last element is cloned, rather than popped. However, I don't know what the consequence of this design will be. I need advice on whether this is a safe change.

There are 4 namespace tests that are failing, but they failed before this change so we haven't gone backwards.

**Comment by Devasta** [comment]
Hi Steve,

So, I have switched this back, the issue is that the vector gets hashmaps added and removed based on the depth down a tree.

If we just clone the last hashmap instead of popping, then the below document gets parsed successfully.

```
<doc>
	<a:child1 xmlns:a="namespace1"/>
	<a:child2/><!-- Should Error -->
</doc>
```

We now have namespace declarations stored as nodes on the source document, but I can add QNames into some other struct if you need?

**Comment by Devasta** [comment]
Right, I can see why this is needed.

I know its very unlikely someone would do it, but if someone had a document like:

```
<a:doc xmlns:a="Namespace1">
	<a:child xmlns:a="Namespace2">
		<a:grandchild xmlns:a="Namespace1"/>
	</a:child>
</a:doc>
```

We wouldn't have the namespaces all together when finished parsing. I think we're going to want to climb up the parent nodes and pull the namespaces from there.

**Comment by ballsteve** [comment]
XML Namespaces are annoying, since you can get pathologically worst case scenarios like that :-(

Climbing the ancestor nodes looks like the way to go.

**Comment by Devasta** [comment]
Issues #98 and #99 are both related to this, if we can get this one resolved fixing both should be straightforward.

**Comment by Devasta** [comment]
Seems this is causing some performance issues on large documents.

If you have a document like:

```
<doc>
    <child>a</child>
    <child>a</child>
    <child>a</child>
    <!-- 1000 rows -->
    <child>a</child>
    <child>a</child>
</doc>
```

Then the namespace vector on the parserstate keeps getting bigger and bigger and is constantly being cloned as the parser tries different functions.

**Comment by ballsteve** [comment]
Hi Daniel,

Sounds like we’ll have to re-think the design for namespaces.

Off the top-of-my-head, perhaps wrap the vector in an Rc so that cloning is cheap? Since the vector must be mutable, use interior mutability?

This is to keep track of in-scope namespaces, so perhaps a different design altogether?

Cheers,
Steve

**Comment by ballsteve** [comment]
Hi Daniel,

I’ve been giving namespaces some thought over the weekend. We need to rationalise and consolidate what has been separate efforts to deal with namespaces.

For that purpose I’ve roughed out a design that deals with both namespaces and QualifiedNames. See attached PDF (which will be included in the xrust documentation).

In addition, I’m also thinking that QualifiedNames might be better treated as interned strings. More on that later.

Cheers,
Steve

￼

**Comment by ballsteve** [comment]
So we're going to have to redesign and reimplement the whole QualifiedName & XML Namespace subsystem. Design documentation has been added in the doc folder (ns_bench branch). This will take a while to sort out...

**Comment by Devasta** [comment]
Hi Steve,

Apologies, just checking on this one: Assuming we implemented it, how does the transformation step then pick up the namespace information for stuff like system-property and custom functions? We've got a system in place now that'll allow us pull together a list of namespaces from the node already by traversing the parents, couldn't we pull them from that?

The vector on the parserstate can be removed to improve performance, but I'm not sure that requires an overhaul this big in scope.

(This is more a question about how construction of the transformation steps works, I've not paid nearly as much attention to that as I should have!)

**Comment by ballsteve** [comment]
Hi Daniel,

The current design has the HashMap storing namespaces in an Rc, i.e. Rc<HashMap<…>>, so that it can be easily, and cheaply, shared. The idea is to add that to the Transform enum. Because it’s an Rc, that’s just a pointer. I haven’t added it to every variant yet, but I haven’t finished the implementation so we’ll see how we go.

If an application is constructing a Transform directly, without a stylesheet document to copy Namespaces from, then we may add a convenience function to map (prefix, namespace URI) pairs to the required structure.

Cheers,
Steve

**Comment by ballsteve** [comment]
Branch issue_95_96 now has the results of some major surgery.

QualifiedName has been changed to use Rc<Value> instead of String. Node names now use Rc<QualifiedName>.
The XML parser interns values and names so that they can be reused.

Running the benchmark results in a 80-90% improvement in the parser(!).

Running the parser on a 100 element document takes, on average, 19.9ms or 0.199ms/element.
Running the parser on a 1000 element document takes, on average, 208ms or 0.2ms/element. This is linear performance.

Previously, the 100 element result was 1.2ms/element and for 1000 elements it was 13.2ms/element. Non-linear performance.

**Comment by ballsteve** [comment]
I've flowed the changes to the rest of the code and pushed the changes in branch issue_95_96.

Not all tests are passing (yet). Once the tests pass, I'll raise a PR to get these changes published. If people are using the latest crate it's going to cause a lot of pain, but better to do that now.

**Comment by ballsteve** [comment]
Daniel, looking at tests parser_config_namespace_nodes_*: these are failing because the assertion expects the root element to have 7 namespace declarations when there are 6. Why do think there should be 7?

**Comment by ballsteve** [comment]
Ah! Using Saxon as a reference implementation, the namespace http://www.w3.org/XML/1998/namespace, prefix "xml" is always declared. I'll add that initialisation to NamespaceMap.

**Comment by Devasta** [comment]
Hi Steve,

Yeah, the namespace spec says that you can but don't need to declare it, I figured explicitly declaring it would be better than having to check for it anywhere it may be used.

Interestingly enough the xmlns namespace can never ever be declared, but I can't seem to find a reason why this would be the case.

**Comment by ballsteve** [comment]
I'm fixing the code by introducing the NamespaceMap object. It's initialisation routine will automatically declare that namespace.

**Comment by ballsteve** [comment]
Phew! I've finished the redesign and implementation of the new namespace design. Just committed and pushed the code in branch issue_95_96. I thought I had to add the 'xml' namespace declaration automatically at document creation, so I introduced the new_document() method. But then I realised that this namespace is always in-scope but is not declared, so I took it out of the document initialisation routine. However, I like the 'new_document()' API (i.e. RNode::new_document()) so I've kept that even though it is now just a convenience for Rc::new(SmiteNode:new()).

I'll now clean up the warnings, add documentation, and then we'll push out a release.
